python-checks/src/test/resources/checks/functionReturnType.py

```diff
@@ -0,0 +1,228 @@
+from typing import List, SupportsFloat, Set, Dict, NoReturn, Text, Generator, Tuple, Union, AnyStr
+
+def builtins():
+  def my_str_nok() -> str:
+    return 42  # Noncompliant
+
+  def my_str_ok() -> str:
+    return "hello"
+
+  def my_int_nok() -> int:
+  #   ^^^^^^^^^^>     ^^^>
+    return "42"  # Noncompliant  {{Return a value of type "int" instead of "str" or update function "my_int_nok" type hint.}}
+  #        ^^^^
+
+  def my_int_ok() -> int:
+    return 42
+
+  def my_none_nok() -> None:
+    return 42  # Noncompliant
+
+  def my_none_ok() -> None:
+    print("hello")
+
+  def my_none_ok_2(param) -> None:
+    if param:
+      return
+    print("hello")
+
+def tuples():
+  def my_tuple_nok() -> int:
+    return 1, 2, 3  # Noncompliant
+
+  def my_tuple_ok() -> tuple:
+    return 1, 2, 3
+
+  def my_tuple_ok_2(param) -> tuple:
+    if param:
+      x = 1, 2, 3
+      return x
+    else:
+      import collections
+      Person = collections.namedtuple('Person', ['name', 'age', 'gender'])
+      return Person(name="Bob", age=30, gender="male")  # OK
+
+  def my_tuple_ok_3(param) -> Union[tuple, bool]:
+    if param:
+      return 1, 2, 3
+    else:
+      return False
+
+  def my_tuple_ok_4(param) -> Union[Tuple["a", "b"], bool]:
+    if param:
+      return 1, 2
+    else:
+      return False
+
+def collections():
+    def my_list_nok() -> List:
+      return 42  # Noncompliant
+
+    def my_list_nok_2() -> List[int]:
+      return 42  # Noncompliant
+
+    def my_list_ok() -> List:
+      return [42]
+
+    def my_set_nok() -> Set:
+      # {} is Empty dict literal
+      return {}  # Noncompliant
+
+    def my_set_nok_2() -> Set[str]:
+      # {} is Empty dict literal
+      return {}  # Noncompliant
+
+    def my_set_ok() -> Set:
+      return {42}  # OK
+
+    def my_dict() -> Dict:
+      return {}  # OK
+
+
+def type_aliases():
+    """We should avoid FPs for type aliases used as type hint"""
+  def my_supports_float() -> SupportsFloat:
+    return 42  # OK
+
+  def my_supports_float() -> SupportsFloat:
+    return "42"  # FN
+
+  def returns_text() -> Text:
+    return "Hello"  # OK
+
+def other_returns():
+  def my_int(param) -> int:
+    if param:
+      return 42
+    else:
+      return "42"  # Noncompliant
+
+  def my_int_returns_none() -> int:
+    return  # Noncompliant
+
+  def my_int_union_type_ok(cond) -> int:
+    if cond:
+      x = 42
+    else:
+      x = "hello"
+    return x  # OK
+
```

Comment:
I think we could add a test case:
```python
  def foo(cond) -> int:
    if cond:
      return 42
    else:
      return "hello"
```

Reply:
There is a test case similar to this one on line 94-98, which does flag the `str` return as `Noncompliant`.

Reply:
ops sorry I missed it